app/controllers/Study.scala

```diff
@@ -307,27 +307,29 @@ final class Study(
   }
 
   private def doImportPgn(id: StudyId, data: StudyForm.importPgn.Data, sri: Sri)(
-      f: List[Chapter] => Result
+      f: (List[Chapter], Option[String]) => Result
   )(using ctx: Context, me: Me): Future[Result] =
     val chapterDatas = data.toChapterDatas
     limit.studyPgnImport(me, rateLimited, cost = chapterDatas.size):
       env.study.api
         .importPgns(id, chapterDatas, sticky = data.sticky, ctx.pref.showRatings)(Who(me, sri))
-        .map(f)
+        .map(f.tupled)
 
   def importPgn(id: StudyId) = AuthBody { ctx ?=> me ?=>
     get("sri").so: sri =>
       bindForm(StudyForm.importPgn.form)(
         doubleJsonFormError,
-        data => doImportPgn(id, data, Sri(sri))(_ => NoContent)
+        data =>
+          doImportPgn(id, data, Sri(sri)): (_, errors) =>
+            errors.fold(NoContent)(NoContent.flashFailure(_))
```

Comment:
flash is for redirecting. Here it's an XHR request so it's not going to work.

The server should return a `BadRequest(jsonError(msg))` and the client should use it for display

Reply:
the client [code](https://github.com/lichess-org/lila/blob/e6040d3ca6b7b05c30ff66fed3426d7107b90899/ui/analyse/src/study/chapterNewForm.ts#L94) is like this:
```typescript
      importPgn(study.data.id, dd).catch(e => {
        if (e.message === 'Too many requests') alert('Limit of 1000 pgn imports every 24 hours');
        throw e;
      });
```
and `importPgn` is an `xhr.text`, which doesn't handle `json` and only returns `Error: 400` for `BadRequest`.

So, we need some refactor to handle this probably. I naively did try to add `if (res.message) throw new Error(res.message);` in [ensureOk](https://github.com/lichess-org/lila/blob/e6040d3ca6b7b05c30ff66fed3426d7107b90899/ui/lib/src/xhr.ts#L16), but didn't work.

I would love some suggestions how to solve this, typescript way.

Reply:
it seems not, as `ensureOk` will catch all exceptions and return different one, so the client will only receive exceptions with `message: Error: 400`.

Reply:
Then use the lower-level `textRaw` to bypass it

https://github.com/lichess-org/lila/blob/master/ui/lib/src/xhr.ts#L48-L48